### PR TITLE
feat(server): i18n for error messages

### PR DESCRIPTION
**Pull request by yk-eukarya**

### CODE REVIEW

File: server/internal/app/graphql.go [hunk 1]
```diff
@@ -81,3 +74,25 @@ func GraphqlAPI(conf GraphQLConfig, dev bool) echo.HandlerFunc {
 		return nil
 	}
 }
+
+func setErrorPresenter(c echo.Context, srv *handler.Server, dev bool) {
```
Comment (rot1024) [inline server/internal/app/graphql.go:78]:
```suggestion
func setGQLErrorPresenter(c echo.Context, srv *handler.Server, dev bool) {
```

File: server/internal/app/i18n.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	bundle = i18n.NewBundle(language.English)
+	bundle.RegisterUnmarshalFunc("yaml", yaml.Unmarshal)
+	_, err := bundle.LoadMessageFileFS(localeFS, "i18n/locale.ja.yaml")
+	if err != nil {
+		log.Error("i18n: failed to load locales")
+	}
+
+}
+
+func I18n(cfg *ServerConfig) echo.MiddlewareFunc {
```
Comment (rot1024) [inline server/internal/app/i18n.go:31]:
```suggestion
func i18nMiddleware(cfg *ServerConfig) echo.MiddlewareFunc {
```

File: server/internal/app/i18n.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			var lang string
+			if op := adapter.Operator(c.Request().Context()); op != nil && op.User != nil {
+				u, err := cfg.Repos.User.FindByID(c.Request().Context(), *op.User)
+				if err == nil && u.Lang() != language.Und {
+					lang = u.Lang().String()
+				}
+			}
+			if lang == "" {
+				lang = c.Request().Header.Get("Accept-Language")
+			}
```
Comment (rot1024) [inline server/internal/app/i18n.go:43]:
```suggestion
			lang := c.Request().Header.Get("Accept-Language")
			if lang == "" {
				if op := adapter.Operator(c.Request().Context()); op != nil && op.User != nil {
					u, err := cfg.Repos.User.FindByID(c.Request().Context(), *op.User)
					if err == nil && u.Lang() != language.Und {
						lang = u.Lang().String()
					}
				} 
			}
```

File: server/internal/app/i18n.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+			l := i18n.NewLocalizer(bundle, lang)
+			c.Set(localizerKey, l)
+			return handlerFunc(c)
+		}
+	}
+}
+
+func t(c echo.Context) *i18n.Localizer {
```
Comment (rot1024) [inline server/internal/app/i18n.go:51]:
```suggestion
func getI18nLocalizer(c echo.Context) *i18n.Localizer {
```

File: server/internal/usecase/interfaces/item.go [hunk 5]
```diff
@@ -4,20 +4,22 @@ import (
...
 	"github.com/reearth/reearth-cms/server/pkg/value"
+	"github.com/reearth/reearthx/rerror"
 	"github.com/reearth/reearthx/usecasex"
 )
 
 var (
 	ErrItemFieldRequired   = errors.New("item field required")
 	ErrInvalidField        = errors.New("invalid field")
-	ErrDuplicatedItemValue = errors.New("duplicated value")
+	ErrDuplicatedItemValue = rerror.NewE(&i18n.Message{ID: "item:duplicated value", Other: "The inserted value is duplicated"})
```
Comment (rot1024) [inline server/internal/usecase/interfaces/item.go:22]:
```suggestion
	ErrDuplicatedItemValue = rerror.NewE(&i18n.Message{ID: "item:duplicated value"})
```

Default value is not needed because locale.en.yaml is used

File: server/tools.go [hunk 6]
```diff
@@ -1,5 +1,8 @@
 //go:build tools
 
+//go:generate go run github.com/nicksnyder/go-i18n/v2/goi18n@latest extract -outdir ./intrnal/app/i18n -format yaml
+//go:generate go run github.com/nicksnyder/go-i18n/v2/goi18n@latest merge -outdir ./intrnal/app/i18n -format yaml ./intrnal/app/i18n/locale.en.yaml ./intrnal/app/i18n/translate.ja.yaml
+
```
Comment (rot1024) [inline server/tools.go:5]:
```suggestion
```

add 

```
_ "github.com/nicksnyder/go-i18n/v2/goi18n"
```

instead if you can

File: server/internal/app/i18n.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/labstack/echo/v4"
+	"github.com/nicksnyder/go-i18n/v2/i18n"
+	"github.com/reearth/reearth-cms/server/internal/adapter"
+	"github.com/reearth/reearthx/log"
+	"golang.org/x/text/language"
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	//go:embed i18n/locale.*.yaml
```
Comment (rot1024) [inline server/internal/app/i18n.go:15]:
```suggestion
	//go:embed ../../i18n/locale.*.yaml
```

**Review by rot1024 — CHANGES_REQUESTED** [review]
Add 2 tasks to run these commands to `Makefile`:

```
go run github.com/nicksnyder/go-i18n/v2/goi18n extract -outdir ./i18n -format yaml
```

```
go run github.com/nicksnyder/go-i18n/v2/goi18n merge -outdir ./i18n -format yaml ./i18n/locale.en.yaml ./i18n/translate.ja.yaml
```

- Do not use `@latest` to prevent unexpected breaking changes
- `i18n` dir is put on the root (`server` dir)

File: server/internal/usecase/interfaces/item.go (hunk 5 above)
Comment (rot1024) [inline server/internal/usecase/interfaces/item.go:22]:
```suggestion
	ErrDuplicatedItemValue = rerror.NewE(&i18n.Message{ID: "duplicated item value", Other: "The inserted value is duplicated"})
```

It is better not to use `:` in a key

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: server/internal/usecase/interfaces/item.go (hunk 5 above)
Comment (yk-eukarya) [reply server/internal/usecase/interfaces/item.go:22]: I was thinking of a way to make the key refer to which model belongs. maybe that's more useful while translating the keys

File: server/internal/app/i18n.go (hunk 7 above)
Comment (yk-eukarya) [reply server/internal/app/i18n.go:15]: `go:embed` does not support parent dir you can check this issue https://github.com/golang/go/issues/46056

File: server/internal/usecase/interfaces/item.go (hunk 5 above)
Comment (yk-eukarya) [reply server/internal/usecase/interfaces/item.go:22]: When I removed the default value the key will not be included in the generated files check this issue: https://github.com/nicksnyder/go-i18n/issues/245

Comment (rot1024) [reply server/internal/usecase/interfaces/item.go:22]: I think It is expected. You can add English messages to YAML, but not needed to add it to codes. You can use ID to recognize what message should be required.

Comment (rot1024) [reply server/internal/usecase/interfaces/item.go:22]: dot is better: `item.duplicated_value`

File: server/internal/app/i18n.go (hunk 7 above)
Comment (rot1024) [reply server/internal/app/i18n.go:15]: I got

File: server/internal/usecase/interfaces/item.go (hunk 5 above)
Comment (yk-eukarya) [reply server/internal/usecase/interfaces/item.go:22]: @rot1024 
But in the current behavior, I could not extract the messages which do not have a translation in the code using the `goi18n extract ...` command

File: server/Makefile [hunk 8]
```diff
@@ -16,4 +16,7 @@ run-db:
 gql:
 	go generate ./internal/adapter/gql
 
-.PHONY: lint test build run-app run-db gql
\ No newline at end of file
+i18n:
+	go run github.com/reearth/reearthx/tools i18n-extract --lang en,ja --outdir i18n ..
```
Comment (rot1024) [inline server/Makefile:20]:
```suggestion
	go run github.com/reearth/reearthx/tools i18n-extract --lang en,ja --outdir i18n .
```

File: server/internal/app/graphql.go [hunk 9]
```diff
@@ -59,20 +60,12 @@ func GraphqlAPI(conf GraphQLConfig, dev bool) echo.HandlerFunc {
...
-			}
-			return graphql.DefaultErrorPresenter(ctx, e)
-		},
-	)
-
 	return func(c echo.Context) error {
 		req := c.Request()
 		ctx := req.Context()
 
+		setGQLErrorPresenter(c, srv, dev)
```
Comment (rot1024) [inline server/internal/app/graphql.go:67]:
```suggestion
		srv.SetErrorPresenter(gqlErrorPresenter(dev))
```

File: server/internal/app/graphql.go [hunk 10]
```diff
@@ -81,3 +74,25 @@ func GraphqlAPI(conf GraphQLConfig, dev bool) echo.HandlerFunc {
 		return nil
 	}
 }
+
+func setGQLErrorPresenter(c echo.Context, srv *handler.Server, dev bool) {
+	srv.SetErrorPresenter(
+		func(ctx context.Context, e error) *gqlerror.Error {
```
Comment (rot1024) [inline server/internal/app/graphql.go:80]:
```suggestion
func gqlErrorPresenter(dev bool) graphql.ErrorPresenterFunc {
	return func(ctx context.Context, e error) *gqlerror.Error {
```

File: server/internal/app/graphql.go [hunk 11]
```diff
@@ -81,3 +74,25 @@ func GraphqlAPI(conf GraphQLConfig, dev bool) echo.HandlerFunc {
...
+			if dev {
+				return gqlerror.ErrorPathf(graphql.GetFieldContext(ctx).Path(), e.Error())
+			}
+
+			if l := getI18nLocalizer(c); l != nil {
```
Comment (rot1024) [inline server/internal/app/graphql.go:85]:
```suggestion
			if l := getI18nLocalizer(ctx); l != nil {
```

File: server/internal/app/i18n.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+			}
+			l := i18n.NewLocalizer(bundle, lang)
+			c.Set(localizerKey, l)
+			return handlerFunc(c)
+		}
+	}
+}
+
+func getI18nLocalizer(c echo.Context) *i18n.Localizer {
+	lz, ok := c.Get(localizerKey).(*i18n.Localizer)
```
Comment (rot1024) [inline server/internal/app/i18n.go:46]:
```suggestion
func getI18nLocalizer(c context.Context) *i18n.Localizer {
	lz, ok := c.Value(localizerKey).(*i18n.Localizer)
```

File: server/internal/app/i18n.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+			if lang == "" {
+				if op := adapter.Operator(c.Request().Context()); op != nil && op.User != nil {
+					u, err := cfg.Repos.User.FindByID(c.Request().Context(), *op.User)
+					if err == nil && u.Lang() != language.Und {
+						lang = u.Lang().String()
+					}
+				}
+			}
+			l := i18n.NewLocalizer(bundle, lang)
+			c.Set(localizerKey, l)
```
Comment (rot1024) [inline server/internal/app/i18n.go:39]:
```suggestion
			ctx = context.WithValue(ctx, localizerKey, l)
			c.SetRequest(req.WithContext(ctx))
```

File: server/internal/app/i18n.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	bundle = i18n.NewBundle(language.English)
+	err := bundle.LoadFS(localeFS, []string{"en.yml", "ja.yml"})
+	if err != nil {
+		log.Error("i18n: failed to load locales")
+	}
+}
+
+func i18nMiddleware(cfg *ServerConfig) echo.MiddlewareFunc {
+	return func(handlerFunc echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
```
Comment (rot1024) [inline server/internal/app/i18n.go:28]:
```suggestion
		return func(c echo.Context) error {
			ctx := c.Request().Context()
```

File: server/internal/app/i18n.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+			lang := c.Request().Header.Get("Accept-Language")
+			if lang == "" {
+				if op := adapter.Operator(c.Request().Context()); op != nil && op.User != nil {
```
Comment (rot1024) [inline server/internal/app/i18n.go:31]:
```suggestion
				if op := adapter.Operator(ctx)); op != nil && op.User != nil {
```

File: server/i18n/ja.yml [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+personal workspace cannot be modified: ""
+projectID is required: ""
+target user does not exist in the workspace: ""
+thread is required: ""
+title cannot be empty: ""
+unsupported entity: ""
+user already joined: ""
+uuid is required: ""
+value is required: ""
+workspace id is required: ""
```
Comment (rot1024) [inline server/i18n/ja.yml:37]: Filling them will be done at another PR, right?

File: server/internal/app/i18n.go [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+	"golang.org/x/text/language"
+)
+
+var (
+	localeFS     = i18nFS.LocalsFS
+	bundle       *i18n.Bundle
+	localizerKey = "localizer"
+)
+
+func init() {
```
Comment (rot1024) [inline server/internal/app/i18n.go:18]:
```suggestion
	localizerKey = localizerKeyType{}
)

type localizerKeyType struct {}

func init() {
```

File: server/internal/app/i18n.go [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"github.com/labstack/echo/v4"
+	i18nFS "github.com/reearth/reearth-cms/server/i18n"
+	"github.com/reearth/reearth-cms/server/internal/adapter"
+	"github.com/reearth/reearthx/i18n"
+	"github.com/reearth/reearthx/log"
+	"golang.org/x/text/language"
+)
+
+var (
```
Comment (rot1024) [inline server/internal/app/i18n.go:12]: Translate https://github.com/reearth/reearthx/blob/main/rerror/error.go#L22 also:

```suggestion
var (
	// rerror
	_ = i18n.T("not found")
	_ = i18n.T("internal")
	_ = i18n.T("invalid params")
	_ = i18n.T("not implemented")

```

**Review by rot1024 — CHANGES_REQUESTED** [review]
Can you add a GitHub Actions step to run `make i18n-ci` to `.github/workflows/ci_server.yml` to check forgotten translations?

File: server/internal/app/i18n.go (hunk 12 above)
Comment (yk-eukarya) [reply server/internal/app/i18n.go:46]: @rot1024 I tried this way, but it does not work

File: server/i18n/ja.yml (hunk 16 above)
Comment (yk-eukarya) [reply server/i18n/ja.yml:37]: I guess Baba san will work on that

File: server/Makefile [hunk 19]
```diff
@@ -16,4 +16,10 @@ run-db:
 gql:
 	go generate ./internal/adapter/gql
 
-.PHONY: lint test build run-app run-db gql
\ No newline at end of file
+i18n:
+	go run github.com/reearth/reearthx/tools i18n-extract --l en,ja --o i18n ..
```
Comment (rot1024) [inline server/Makefile:20]:
```suggestion
	go run github.com/reearth/reearthx/tools i18n-extract -l en,ja -o i18n .
```

File: server/Makefile [hunk 20]
```diff
@@ -16,4 +16,10 @@ run-db:
...
+
+i18n-ci:
+	go run github.com/reearth/reearthx/tools i18n-extract --l en,ja --o i18n --fail-on-update ..
```
Comment (rot1024) [inline server/Makefile:23]:
```suggestion
	go run github.com/reearth/reearthx/tools i18n-extract -l en,ja -o i18n --fail-on-update .
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: server/Makefile (hunk 20 above)
Comment (rot1024) [inline server/Makefile:23]:
```suggestion
i18n:
	go run github.com/reearth/reearthx/tools i18n-extract -l en,ja -o i18n .

i18n-ci:
	go run github.com/reearth/reearthx/tools i18n-extract -l en,ja -o i18n --fail-on-update .
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

**Review by rot1024 — APPROVED** [review]

File: server/i18n/ja.yml [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+invalid base URL: ""
+invalid cursor: ""
+invalid default values: ""
+invalid document: ""
+invalid email: ""
+invalid email address: ""
+invalid email or password: ""
+invalid field: ""
+invalid file: ""
+invalid iss: ""
```
Comment (HideBa) [inline server/i18n/ja.yml:47]:
```suggestion
Comment already exist in this thread: コメントはすでにこのスレッドに存在します。
Comment does not exist in this thread: コメントはこのスレッドに存在しません。
already locked: 既にロック済みです。
already published: 既に公開済みです。
archived: アーカイブ済み
auth0 is not set up: Auth0が設定されていません。
"auth0: domain is not set": Auth0のドメインが設定されていません。
bucket name is empty: ストレージバケット名が空白です。
can't update by approve: 
cannot change the role of the workspace owner: ワークスペースのオーナーのロールを変更することはできません。
cannot delete workspace because at least one project is left: プロジェクトが存在するためワークスペースを削除することはできません。
comment already exist in this thread: コメントは既にこのスレッドに存在します。
comment does not exist in this thread: コメントはこのスレッドに存在しません。
comment not found: コメントが見つかりませんでした。
could not get user info: ユーザー情報が取得できませんでした。
createdBy is required: createdByは必須です。
duplicated key: キーが重複しています。
duplicated user: ユーザーが重複しています。
duplicated value: 値が重複すています。
encoding password: パスワードのエンコーディング
failed to auth: 認証に失敗しました。
failed to create asset: アセットの作成に失敗しました。
failed to delete file: ファイルの削除に失敗しました。
failed to lock: ロックに失敗しました。
failed to update user: ユーザー情報の更新に失敗しました。
failed to upload file: ファイルのアップロードに失敗しました。
field not found: ファイルが見つかりませんでした。
field value exist: フィールドの値はすでに存在します。
file is required: ファイルは必須です。
file not found: ファイルが見つかりませんでした。
file not included: ファイルが含まれていません。
file size cannot be zero: ファイルサイズは0以下にできません。
file too large: ファイルサイズが大きすぎます。
internal: 内部
invalid URL: 無効なURLです。
invalid access token: 向こうなアクセストークンです。
invalid alias: 無効なエイリアスです。
invalid base URL: 無効なベースURLです。
invalid cursor: 無効なカーソルです。
invalid default values: 無効なデフォルト値です。
invalid document: 無効なドキュメントです。
invalid email: 無効なEmailです。
invalid email address: 無効なEmailアドレスです。
invalid email or password: 無効なEmailもしくはパスワードです。
invalid field: 無効なフィールドです。
invalid file: 無効なファイルです。
invalid iss: 無効なissです。
```

File: server/i18n/ja.yml [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+target workspace still has some project: ""
+thread is required: ""
+title cannot be empty: ""
+unauthorized: ""
+unsupported entity: ""
+user already exists: ""
+user already joined: ""
+uuid is required: ""
+value is required: ""
+workspace id is required: ""
```
Comment (HideBa) [inline server/i18n/ja.yml:104]:
```suggestion
invalid issuer: 無効なissuerです。
invalid key: 無効なキーです。
invalid lang: 無効な言語です。
invalid object: 無効なオブジェクトです。
invalid operator: 無効なオペレーターです。
invalid parameters: 無効なパラメータです。
invalid params: 無効なパラメーターです。
invalid password: 無効なパスワードです。
invalid password confirmation: パスワードが一致しません。
invalid password reset request: 無効なリセットリクエストです。
invalid project: 無効なプロジェクトです。
invalid role: 無効なロールです。
invalid secret: 無効なシークレットです。
invalid smtp url: 無効なSMTP URLです。
invalid type: 無効な型です。
invalid type property: 無効な型プロパティです。
invalid user id: 無効なユーザーIDです。
invalid user name: 無効なユーザー名です。
invalid value: 無効な値です。
invalid values: 無効な値です。
invalid workspace name: 無効なワークスペース名です。
item field required: このフィールドは必須項目です。
items cannot be empty: アイテムは空にできません。
max must be larger then min: 最大値は最小値より大きい必要があります。
missing fields: フィールドが不足しています。
missing required config: 必須項目が設定されていません。
model key is already used by another model: このキーはすでに別のモデルで使用されています。
not found: 見つかりませんでした。
not implemented: 未実装です。
not implemented yet: 未実装です。
not locked: ロックされていません。
not verified user: 未認証のユーザーです。
nothing is updated: アップデートされた項目はありません。
only requests with status waiting can be approved: レビュー待ちのリクエストのみ承認可能です。
only reviewers can approve: レビュワーのみ承認可能です。
operation denied: 操作が拒否されました。
owner user cannot leave from the workspace: オーナーはワークスペースを抜けることができません。
password at least 8 characters: パスワードは最低８文字必要です。
password should have lower case letters: パスワードには小文字を含める必要があります。
password should have numbers: パスワードには数字を含める必要があります。
password should have upper case letters: パスワードには大文字を含める必要があります。
personal workspace cannot be modified: パーソナルワークスペースは変更できません。
project alias is already used by another project: プロジェクトエイリアスはすでに別のプロジェクトで使用されています。
project alias is not set: プロジェクトエイリアスが設定されていません。
projectID is required: プロジェクトIDは必須です。
reviewer should be owner or maintainer: レビュワーはオーナーもしくはメインテイナーである必要があります。
target user does not exist in the workspace: 対象のユーザーはワークスペースに存在しません。
target workspace still has some project: 対象のワークスペースにプロジェクトが存在します。
thread is required: スレッドは必須です。
title cannot be empty: タイトルは必須です。
unauthorized: 未認証
unsupported entity: サポートされていないエンティティ
user already exists: ユーザーはすでに存在します。
user already joined: ユーザーはすでに参加しています。
uuid is required: UUIDは必須です。
value is required: 値は必須です。
workspace id is required: ワークスペースIDは必須です。
```

**Review by HideBa — APPROVED** [review]
Kindly check it!
